pkg/volume/csi/csi_attacher.go

```diff
@@ -324,7 +325,8 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 		return err
 	}
 	defer func() {
-		if err != nil {
+		// Only for non-timedout errors remove the mount directory
+		if err != nil && !volumetypes.IsOperationTimeOutError(err) {
```

Comment:
It is possible to fix but this, it is kinda super ugly...

One can for example - return two errors from `MountDevice` call. One for CSI error and another for non-csi error. Both errors get recorded somewhere(asow or dsow not sure yet) and when next `MountDevice` call is made then both errors are supplied as arguments along with other args. So we can compare last csi error and current csi error and determine if operation is pending and not remove the directory for non-csi errors (and if both current and last csi-error is timeout).

It is no silver bullet however, I am afraid this will lead to orphan pod/volume directories left on the nodes and we have to carefully code this. There might be other unforeseen corner cases lurking around which we haven't thought through if we take this route.

Right now - my 2 cents are on calling current fix "good enough" and see if we receive bug reports from people about staging directories disappearing. I am not sure added complexity of tracking additional states is worth the cost. :(

Reply:
In external-provisioner we solved it by returning three values from [`ProvisionExt()`](https://github.com/jsafrane/sig-storage-lib-external-provisioner/blob/master/controller/volume.go#L76):

- `pv` - unrelated to NodeStage/Publish
- `err`: the error itself, mostly for logging
- `ProvisioningState` - state of the provisioning (when `err != nil`)
  - `Background` - CSI provisioner has reached CSI driver and got non-final error that means the provisioning may be in progress (e.g. timeout)
  - `Finished` - CSI provisioner has reached CSI driver and got final error that means the provisioning is for sure not in progress
  - `NoChange` - the provisioner did not reach the CSI driver (e.g. failed to get Secrets from API server). CSI provisioner has no clue in what state the provisioning is, but it is in the same state as before the call.

Caller of `ProvisionExt` (i.e. the provisioning lib) then keeps a list of provisionings in progress. IMO it's equal to uncertain state introduced in this PR.

An additional return val in `SetUp` / `MountDevice` ( / `TearDown` / `UnmountDevice`) could be enough, where the caller (`*Operation`) would mark the volume as uncertain when it receives `Background`, mark it as unmounted after `Finished` and leave the volume as it was before after `NoChange`.

Reply:
fixed